lib/core/include/scipp/core/spatial_transforms.h

```diff
@@ -0,0 +1,174 @@
+#include <Eigen/Geometry>
+
+#include "scipp/core/dtype.h"
+
+#include <type_traits>
+#include <optional>
+
+namespace scipp::core {
+
+class Rotation {
+private:
+  // Store as quaterniond as this is more space efficient than storing as matrix
+  // (4 doubles for quat vs 9 doubles for 3x3 matrix).
+  Eigen::Quaterniond m_quat;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  Rotation() : m_quat(Eigen::Quaterniond::Identity()){};
+  explicit Rotation(const Eigen::Quaterniond &x) : m_quat(x){};
+
+  [[nodiscard]] Eigen::Quaterniond quat() const {
+    return m_quat;
+  }
+
+  bool operator==(const Rotation &other) const {
+    return m_quat.w() == other.m_quat.w() && 
+        m_quat.x() == other.m_quat.x() && 
+        m_quat.y() == other.m_quat.y() && 
+        m_quat.z() == other.m_quat.z();
+  }
+
+  const double get(const int index) const {
+      if (index == 0)
+          return m_quat.w();
+      else if (index == 1)
+          return m_quat.x();
+      else if (index == 2)
+          return m_quat.y();
+      else if (index == 3)
+          return m_quat.z();
+      else
+          throw std::out_of_range("invalid index in get for Rotation");
+  }
+};
+
+class Scaling {
+private:
+  Eigen::DiagonalMatrix<double, 3> m_mat;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  Scaling()
+      : m_mat(Eigen::Matrix3d::Identity().diagonal().asDiagonal()){};
+  explicit Scaling(const Eigen::DiagonalMatrix<double, 3> &x)
+      : m_mat(x){};
+
+  [[nodiscard]] Eigen::Matrix3d matrix() const { return m_mat; }
+
+  bool operator==(const Scaling &other) const {
+    return m_mat.diagonal() == other.m_mat.diagonal();
+  }
+
+  const double get(const int index) const {
+      return m_mat.diagonal()(index);
+  }
+};
+
+class Translation {
+private:
+  Eigen::Vector3d m_vec;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  Translation() : m_vec(Eigen::Vector3d(0, 0, 0)){};
+  explicit Translation(const Eigen::Vector3d &x) : m_vec(x){};
+
+  [[nodiscard]] Eigen::Vector3d vector() const { return m_vec; }
+
+  bool operator==(const Translation &other) const {
+    return m_vec == other.m_vec;
+  }
+
+  const double get(const int index) const {
+      return m_vec(index);
+  }
+};
+
+template<class T> struct asEigenType_t { using type = T; };
+template<> struct asEigenType_t<Rotation> { using type = Eigen::Quaterniond; };
+template<> struct asEigenType_t<Scaling> { using type = Eigen::Matrix3d; };
+template<> struct asEigenType_t<Translation> { using type = Eigen::Translation<double, 3>; };
+
+template<typename T, typename R = asEigenType_t<T>::type> inline const R asEigenType(const T &obj) { return obj; };
+template<> inline const Eigen::Quaterniond asEigenType(const Rotation &obj) { return obj.quat(); }
+template<> inline const Eigen::Matrix3d asEigenType(const Scaling &obj) { return obj.matrix(); }
+template<> inline const Eigen::Translation<double, 3> asEigenType(const Translation &obj) { return Eigen::Translation<double, 3>(obj.vector()); }
```

Comment:
Should probably not return by value in any of the above?

Reply:
I think at least `Translation` needs to return by value, as it is constructing an `Eigen::Translation` object which would be out of scope if we return it by reference? This seems to be what other people online suggest anyway with eigen types - return by value and let the compiler do any return-value optimisations needed to avoid copies etc. The [eigen docs](https://eigen.tuxfamily.org/dox/group__TopicPassingByValue.html) are also fairly explicit that `...there is no problem with functions that return objects by value. `

Having said that, I did try to get the other template specializations to return by ref *anyway*, but couldn't quite figure out the right syntax for it... if you feel strongly about this could I have a hand with it please?

Reply:
> I think at least Translation needs to return by value, as it is constructing an Eigen::Translation object which would be out of scope if we return it by reference?

Yes, since the translation is not stored in the wrapper. I suppose `Translation` does not actually copy the values?

> The eigen docs are also fairly explicit that ...there is no problem with functions that return objects by value. 

There is no *problem* with it, but that does not mean that it is optimal for performance.

> Having said that, I did try to get the other template specializations to return by ref anyway, but couldn't quite figure out the right syntax for it... if you feel strongly about this could I have a hand with it please?

I think we have to do this properly, yes. If you return by value in the wrong place, you end up with expression templates referencing things on the stack that are not in scope any more.